test/Controller/SubmitControllerTest.php

```diff
@@ -92,6 +92,66 @@ public function it_redirects_you_to_xpub_with_a_jwt()
         $this->assertFalse($jwt['new-session']);
     }
 
+    /**
+     * @test
+     */
+    public function it_redirects_you_to_a_trusted_url_with_a_jwt()
+    {
+        $client = static::createClient();
+        $this->logIn($client);
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://localhost/path?query=arg'));
+        $response = $client->getResponse();
+
+        $this->assertTrue($response->isRedirect());
+        $location = new Uri($response->headers->get('Location'));
+
+        $this->assertSameUri('http://localhost/path?query=arg', $location->withFragment(''));
+
+        $jwt = (array) JWT::decode($location->getFragment(), $this->getParameter('xpub_client_secret'), ['HS256']);
+
+        $this->assertFalse($jwt['new-session']);
+    }
+
+    /**
+     * @test
+     */
+    public function it_redirects_you_to_a_trusted_url_with_a_jwt_in_query_argument()
+    {
+        $client = static::createClient();
+        $this->logIn($client);
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://localhost/path?query=arg').'&token_in_query=true');
+        $response = $client->getResponse();
+
+        $this->assertTrue($response->isRedirect());
+        $location = new Uri($response->headers->get('Location'));
+
+        $this->assertUriStartsWith('http://localhost/path?query=arg', $location->withFragment(''));
```

Comment:
Can probably unset the `token` parameter and use `assertSameUri()`?

Reply:
done